File: Perl/euler-485-maximum-number-of-divisors.pl
:- use_module(library(clpfd)).

% Sieve of Eratosthenes for generating primes
sieve(Max, Primes) :-
    numlist(2, Max, NumList),
    sieve(NumList, [], RevPrimes),
    reverse(RevPrimes, Primes).

sieve([], Primes, Primes).
sieve([P|Rs], Acc, Primes) :-
    exclude(divisible_by(P), Rs, Filtered),
    sieve(Filtered, [P|Acc], Primes).

divisible_by(P, N) :-
    N mod P #= 0.

% Prime factorization
prime_factors(N, Factors) :-
    Max is floor(sqrt(N)),
    sieve(Max, Primes),
    factorize(N, Primes, Factors).

factorize(1, _, []).
factorize(N, [P|Ps], Factors) :-
    (   N mod P #= 0
    ->  N1 #= N // P,
        factorize(N1, [P|Ps], F),
        Factors = [P|F]
    ;   factorize(N, Ps, Factors)
    ).

% Count divisors from prime factors
count_divisors(Factors, Count) :-
    run_length_encode(Factors, Encoded),
    findall(C, (
        member(_-M, Encoded),
        C #= M + 1
    ), Counts),
    foldl(mult, Counts, 1, Count).

mult(A, B, Result) :-
    Result #= A * B.

% Run-length encoding for factors
run_length_encode(List, Encoded) :-
    pack(List, Packed),
    maplist(length, Packed, Lengths),
    pairs_keys_values(Encoded, Packed, Lengths).

pack([], []).
pack([X|Xs], [[X|Ys]|Zs]) :-
    span(X, Xs, Ys, Ws),
    pack(Ws, Zs).

span(X, [X|Xs], [X|Ys], Zs) :-
    !, span(X, Xs, Ys, Zs).
span(_, Xs, [], Xs).

% Pre-calculate divisors for numbers up to MaxValue
precalculate_divisors(MaxValue, Divisors) :-
    findall(DivCount, (between(1, MaxValue, Num), d(Num, DivCount)), Divisors).

% d function (number of divisors)
d(N, DivisorCount) :-
    prime_factors(N, Factors),
    count_divisors(Factors, DivisorCount).

% M function optimized with precalculated divisors
m(N, K, MaxDivisor, Precalculated) :-
    Upper is N + K - 1,
    findall(DivCount, (between(N, Upper, Num), nth1(Num, Precalculated, DivCount)), DivCounts),
    max_list(DivCounts, MaxDivisor).

% S function optimized with precalculated divisors
s(U, K, Sum, Precalculated) :-
    Upper is U - K + 1,
    findall(MaxD, (between(1, Upper, N), m(N, K, MaxD, Precalculated)), MaxDs),
    sum_list(MaxDs, Sum).

% Main entry point
main :-
    precalculate_divisors(1000, PrecalculatedDivisors),
    s(1000, 10, Result, PrecalculatedDivisors),
    format("S(1000, 10) = ~w~n", [Result]).

:- initialization(main).

File: Haskell/euler-485-maximum-number-of-divisors.hs
import Data.List (group)
import System.Environment (getArgs)
import qualified Data.Map as Map


-- Sieve of Eratosthenes to generate primes
primes :: [Integer]
primes = 2 : filter (isPrime primes) [3,5..]
  where
    isPrime (p:ps) n = p*p > n || n `mod` p /= 0 && isPrime ps n

-- Function to calculate the number of divisors of n
d :: Integer -> Map.Map Integer Integer -> (Integer, Map.Map Integer Integer)
d n cache =
  case Map.lookup n cache of
    Just res -> (res, cache)
    Nothing  -> let res = divisorCount n
                in (res, Map.insert n res cache)
  where
    divisorCount n = product . map ((+1) . fromIntegral . length) . group . primeFactors $ n
    primeFactors n = factorize n primes
    factorize n (p:ps) 
      | p*p > n    = [n]
      | n `mod` p == 0 = p : factorize (n `div` p) (p:ps)
      | otherwise  = factorize n ps

-- Function M(n, k)
m :: Integer -> Integer -> Map.Map Integer Integer -> (Integer, Map.Map Integer Integer)
m n k cache = foldl (\(maxD, c) j -> let (dj, c') = d j c in (max maxD dj, c')) (0, cache) [n..n+k-1]

-- Function S(u, k)
s :: Integer -> Integer -> Integer
s u k = fst $ foldl (\(sumM, c) n -> let (mn, c') = m n k c in (sumM + mn, c')) (0, Map.empty) [1..u-k+1]

-- Main function to accept command line arguments
main :: IO ()
main = do
    args <- getArgs
    let u = read (args !! 0) :: Integer
    let k = read (args !! 1) :: Integer
    print $ s u k